Add Rollout Process (#1178)
sourcegraph/about: handbook/product/rollout_process.md

# Rollout Process

The rollout process consists of *Pre-Launch* phase and *Deployment* phase. We can push changes to Sourcegraph.com when we
want as we own the deployment process. For enterprise instances of Sourcegraph.com, we push a new release monthly and our
clients decode to upgrade to the latest release when they want.

### Sourcegraph.com

#### Pre-Launch
1. User testing
- Hallway test with internal users.

> - Open the feature to internal users by turning it on for the Sourcegraph organization. Make sure it's opened with enough

Author: I linked @slimsag's doc which includes information on how to update both the site settings and global settings. It needs to be updated to add a bit more context for new teammates.

> - Open the feature to internal users by turning it on for the Sourcegraph organization. Make sure it's opened with enough

Reviewer: Agree to include a link - but org settings don't need a PR, they are managed here: https://sourcegraph.com/organizations/sourcegraph/settings

# Rollout Process

The rollout process consists of *Pre-Launch* phase and *Deployment* phase. We can push changes to Sourcegraph.com when we
want as we own the deployment process. For enterprise instances of Sourcegraph.com, we push a new release monthly and our
clients decide to upgrade to the latest release when they want.

### Sourcegraph.com

#### Pre-Launch
1. User Testing 
- Hallway test with internal users.
- Open the feature to internal users by turning it on for the Sourcegraph organization. Make sure it's opened with enough
time left before planned launch date to receive and address all the feedback received. 
2. Design 
- Test the feature for both themes, Light and Dark. Check for icon consistency across themes.
* Test on small, medium, large and extra-large screen sizes.
  * Small screens are important to consider for a good experience when the window is resized or in side-by-side mode.
* Code examples in design
  * Font should be "code font" or monospace.

>   * Easy to copy and paste (no fancy quotes).

Reviewer 2: May want to add that these are only applicable for UI-facing changes (or alternatively, that this document lays out the process for rolling out UI-facing launches)

Author: I don't know what rolling out backend changes would look like. For now, followed your above comment and added a line to skip if not applicable.

# Feature rollout process

Features come in many different sizes and shapes, and the process for introducing new functionality ranges with these differences. For large or significantly impactful changes or changes that simply need a bit more time to bake, it is encouraged that the following rollout process is followed.

## Sourcegraph Cloud

> Sourcegraph Cloud is continuously deployed with all new updates to master. We maintain a [releasability contract](../engineering/continuous_releasability.md) and require all new features to be released behind a feature flag to ensure that functionality can be turned off if a problem arises.

Reviewer 3: I don't think it's good to reference the releasability contract here. For example, [private code CTAs](https://github.com/sourcegraph/sourcegraph/pull/12288#discussion_r456694318) shouldn't be feature flagged, IMO, but it seems that product does want to discuss/approve this before merging (https://sourcegraph.slack.com/archives/C0C324C91/p1595009322049800). Should we have two different paths for features that should and shouldn't be feature flagged?

Author: @attfarhan, curious why think it shouldn't be feature flagged. Is it because it's primarily a UI change that can be fixed easily (instead of reverted?). I can add exceptions we decide on to the releasability contract in a follow-up PR for feedback.
Private code CTA was behind a master feature (`showRepogroupHomepage`) with one non-ideal way to turn it off if a problem arises. To your point, we don't have a section for how incremental changes should be rolled out. Would that be helpful in this doc?

Reviewer 3: Yeah I'm basing it off the fact that we _didn't_ feature flag it. It was an incremental change, as you said. But yeah you're right that it was behind an umbrella feature flag. I think guidelines for incremental changes could be helpful.

### Before merge

- Run hallway tests with internal users
- Complete a final [design review](design/design_process.md#final-review)
- Review documentation
- Review analytics and ensure desired metrics have been added to the feature

> - Confirm feature flag functionality

Reviewer 3: Do we need to add a step here where the product team must approve the PR before merging? I worry that the code will be ready, the PR is approved by other developers, but the product team doesn't approve yet. It then becomes unclear when the dev can feel confident merging.

Reviewer 4: Seems like all of this can actually be after merge and deployment if things are feature flagged?

# Rollout Process

The rollout process consists of *Pre-Launch* phase and *Deployment* phase. We can push changes to Sourcegraph.com when we
want as we own the deployment process. For enterprise instances of Sourcegraph.com, we push a new release monthly and our
clients decide to upgrade to the latest release when they want.

## Sourcegraph.com

> ### Pre-Launch
> **User Testing** 
> - Hallway test with internal users.
> - Open the feature to internal users by turning it on for the Sourcegraph organization. Make sure it's opened with enough
> time left before planned launch date to receive and address all the feedback received.  Go to [our Sourcegraph organization settings](https://sourcegraph.com/organizations/sourcegraph/settings) to enable your feature only for internal users in the Sourcegraph organization.
> 
> **Design** 
> (skip if it doesn't apply)
> - Test the feature for both themes, Light and Dark. Check for icon consistency across themes.
> * Test on small, medium, large and extra-large screen sizes.
>   * Small screens are important to consider for a good experience when the window is resized or in side-by-side mode.
> * Code examples in design
>   * Font should be "code font" or monospace.
>   * Easy to copy and paste (no fancy quotes).
> 
> **Analytics** 
> - Add and test logging for critical flows.
> 
> **Approvals**
> - Receive approval from engineering, design, data, product and other stakeholders before launch. 
> 
> **Bug Tracking**
> - Track and ensure high priority bugs in GitHub issues. Ensure they are closed before launch.
> - Track all lower priority bugs that have to be fixed soon.

Reviewer 5: ```suggestion
### Before merge

- Run hallway tests with internal users
- Complete a final [design review](design/design_process.md#final-review)
- Review documentation
- Review analytics and ensure desired metrics have been added to the feature
- Confirm feature flag functionality

### After merge, before launch

1. **Gather internal feedback:** Enable the feature flag in the [Sourcegraph organization settings](https://sourcegraph.com/organizations/sourcegraph/settings) to enable your feature for all Sourcegraph team members. Be sure to leave enough time for folks to experience the feature in their workflows and provide feedback.
1. **Analytics:** Validate logging is working for critical flows
1. **Approvals:** Recieve approval from key stakeholders.
1. **Bug Tracking:** Keep track of all feedback.
   - Track and ensure high priority bugs in GitHub issues. Ensure they are closed before launch.
   - Track all lower priority bugs that have to be fixed soon.
```

Reviewer 5: Link to updated final design review in PR: https://github.com/sourcegraph/about/pull/1225

> ### Deployment
> **Push to Sourcegraph.com**
> - Follow the steps in this [document](https://about.sourcegraph.com/handbook/engineering/distribution/update_sourcegraph_website) to enable your feature in global settings and to push it to all users on Sourcegraph.com.
> - In the PR that pushes changes live, add everyone who gave appproval for launch as reviewers.
> 
> **Metrics**
> - Share analytics for monitoring the feature shipped. Track metrics for regressions. 

Reviewer 5: ```suggestion
### Launch

1. **Enable for all to Sourcegraph Cloud users**
   - Follow the steps in this [document](https://about.sourcegraph.com/handbook/engineering/distribution/update_sourcegraph_website) to enable your feature in global settings and to push it to all users on Sourcegraph.com.
   - In the PR that pushes changes live, add everyone who gave appproval for launch as reviewers.
1. **Metrics**
   - Share analytics for monitoring the feature shipped. Track metrics for regressions. 
```

> ### Post-Launch
> **Marketing**
> - Share update with marketing.

Reviewer 5: ```suggestion
### Post-Launch

1. **Marketing:** Share update with marketing.
1. **Metrics:** continue to track metrics to ensure expected outcomes are achieved.
1. **Enable by default:** Once a feature has been validated on Sourcegraph Cloud it can be enabled by default to be included in the next release. The feature flag should be left in place for at least one release cycle so that customers can disable the feature if problems arise.
```

# Rollout Process

The rollout process consists of *Pre-Launch* phase and *Deployment* phase. We can push changes to Sourcegraph.com when we
want as we own the deployment process. For enterprise instances of Sourcegraph.com, we push a new release monthly and our
clients decide to upgrade to the latest release when they want.

## Sourcegraph.com

### Pre-Launch
**User Testing** 
- Hallway test with internal users.
- Open the feature to internal users by turning it on for the Sourcegraph organization. Make sure it's opened with enough

> time left before planned launch date to receive and address all the feedback received.  Follow this [document](https://about.sourcegraph.com/handbook/engineering/distribution/update_sourcegraph_website) to enable your feature in site settings and turn it on only for internal users in the Sourcegraph organization. TODO: Update doc linked above to add a bit more content for new teammates

Reviewer: Turning it on in site settings turns it on for everyone, it needs to be set in org settings.
```suggestion
time left before planned launch date to receive and address all the feedback received.  Go to [our Sourcegraph organization settings](https://sourcegraph.com/organizations/sourcegraph/settings) to enable your feature only for internal users in the Sourcegraph organization.
```

Author: Added a test search scope to test this quickly myself too. Thanks for fixing this!

**Design** 
(skip if it doesn't apply)
- Test the feature for both themes, Light and Dark. Check for icon consistency across themes.
* Test on small, medium, large and extra-large screen sizes.
  * Small screens are important to consider for a good experience when the window is resized or in side-by-side mode.
* Code examples in design
  * Font should be "code font" or monospace.

>   * Easy to copy and paste (no fancy quotes).

Reviewer: I think these things belong more into a style guide than into a rollout process

Reviewer 6: Agreed, all of this should have been caught in design review.

Author: Agree, I'd like to link to a style guide here eventually. I see this style guide living in `/design`.
